### ISSUE TITLE
New Component: InputCurrency

### ISSUE THREAD

**Issue by melloware** [primefaces-extensions/primefaces-extensions.github.com#409]
I am working on a new component called InputCurrency that is based on JQuery MaskMoney http://plentz.github.io/jquery-maskmoney/

I think this fills a need that InputNumber does not as it allows for ATM or Calculator like data entry where you can type "1" three times you get 0.01, 0.11, 1.11 just like a calculator or ATM.  It also allows for String submittal to a backing bean so if you wanted "$ 1.11 USD" the value could be sent as a string.

I have it all working an showcase examples updated.  @tandraschko any objection to me checking it in?  We can always remove it if the team does not approve but I find it handy for quick entry of prices in my webapps.

**Comment by tandraschko** [comment]
+0
we are actually in maintenance mode :D

Also not sure about the plugin. I looks ok but e.g. it's not possible to mark only a part of the currency.
We also created a InputCurrency in our application which is derived from InputNumber. As you said, there's a missing feature.

**Comment by ova2** [comment]
I agree with Thomas. I'm not active in this project anymore.

**Comment by melloware** [comment]
While I understand completely your situation, this is unfortunate.  

PF Extensions is where the PF Community goes to look for components that are missing in the PF universe.  That is what brought me here as our projects heavily use Timepicker and CodeMirror.  Previously we were using InputNumber and KeyFilter that have since graduated into PrimeFaces core.

By not taking new component submissions and without this vital project you are only leaving PF community with a few choices when it comes to components that don't exist:

1. Hope PF Core implements the component they want or need.
2. Create a separate github page housing that single JSF component that won't really get adopted because no one will know about it or trust it.
3. Or they write it in their own in-house project never bothering to share it with the world because its too much trouble to start up a new GitHub project for a single JSF component.

Other useful components I was considering creating were:

- Calculator: http://keith-wood.name/calculator.html
- TabSlideout: https://github.com/hawk-ip/jquery.tabSlideOut.js/tree/master

**Comment by tandraschko** [comment]
I understand you point of view, too.
The problem is, we tried many times to get bugfixes from the community or get new active commiters.
The result was, that me and oleg were mainly the only maintainers here.
Oleg doesn't JSF anymore, if i would create new components, they were likely end in PF directly.
I currently don't even use a extensions component anymore, some private customers use 3-4 components of extensions.

The next problem is, that Oleg is currently the only one who can do releases.

Personally i'm not 100% against a new component, but i'm -0,5 of creating components which are already very similar available, which are really really rarely used or components which can be done easily done in customers in simple composites.

**Comment by ova2** [comment]
I'm still working with JSF, but much less than before and I don't use Extensions (except pe:remoteCommand). I only can do releases as before. Thomas is working for PrimeFaces core. We only accept Pull Requests. If you want, we can add you as commiter, then you can commit directly. No problem.

**Comment by melloware** [comment]
@ova2 Thomas made me a committer so I can maintain CodeMirror and help out any way I can.  I can't thank the both of you enough for starting and maintaining this great effort. And I have no problem where you two are not actively fixing issues and only accepting pull requests.  

I plan on reviewing bugs and helping out any way I can.   I just felt that if even 1 other person out there found a component useful that adding new components and keeping this project alive is well worth it.  

So it sounds like Oleg is OK with new components but Thomas is not?

Also Oleg as long as you don't mind doing the releases that is great but if you ever got tired of it is there a way for you to turn the release process over to someone else?  Or is there an issue with it being tied to your PGP key for Sonatype etc?

**Comment by Rapster** [comment]
I have to agree with @melloware, the project have to continue to evolve, bring new cool components, it is sad how commits in PF or PE have been so low these last months... Regarding inputCurrency, I think it's a nice and handy component, and I encourage @melloware to commit it. @trandraschko leave the implementation up to your client and as a composite is a bad idea, first, most of them don't know much about JSF, because of the deep learning  curve, secondly, composite in JSF are tricky to use, thirdly, as soon it is used by two client, make it as an official and reusable component :)

Regarding those two components @melloware, I'm not sure about the usefulness of tabSlideOut?

**Comment by ova2** [comment]
Hi guys, melloware and Rapster. I'm very happy that you are on board! Sure, you can add new components and keep the project alive. I hope Thomas is fine with that :-)

Yes, the releases are tied to my PGP key for Sonatype. Can we change this somehow? By the way, for the first release I've created this issue at Sonatype https://issues.sonatype.org/browse/OSSRH-2317 You can ask questions and report release related bugs there. Last time I could not do releases by Maven release plugin (don't know why) and had to create JAR bundlers manually and deploy them to Sonatype. What are your e-mails? I can send you my account for Sonatype per e-mail (you know, you have to log in and push the release button there when the release has been done).

**Comment by tandraschko** [comment]
I'm -0,5 of InputCurrency because its very similar (~95%) to InputNumber. There is only the special input handling like you mentioned in the first post but this also bring problems, like that you can't mark text anymore and replace it.

Calculator: Whats a usecase? I develop many financal applications and we never needed something
tabSlideOut: Looks ok, but should be per default better styled.

I'm not -1 for every component but we should really look forward to only put in great and usefull components.

**Comment by melloware** [comment]
@ova2 Email is info@melloware.com.  Basically from reading that bug report we have to update our pom's to not use the old way the releases used to work for Sonatype.  I might be able to take a crack at it.

@tandraschko @Rapster  InputCurrency you are right it comes with some flaws like you mentioned but if someone truly just wants ATM like input i think they would live with those flaws.  If InputNumber had a feature that was something like type="calculator" that emulated this behavior then I wouldn't have needed to create InputCurrency to meet the need.  My users complain if they mouse click in InputNumber they just punch their numbers with their head down  and they look up to the screen to see it didn't take their entry as expected.  

Calculator: You are right I could argue the use case is only valid for Quickbooks/Mint type checkbook applications or double entry accounting.  I could live without this just thought it was cool and was already themed for Jquery/Themeroller etc.

TabSlideOut: We do use this for our customers to provide quick feedback that "sticks" to the screen as the scroll.  A great demo is here:  https://dashboard.hawkip.com/gateway/login  Click on his TabSlideout's to get a feel for how useful it can be.  Maybe this is accomplished by some other PF component but nothing that I found that allowed it to stick to the side of the screen and also allow it to be sticky or not?

**Comment by tandraschko** [comment]
Feel free to commit Calculator + TabSlideOut (but don't merge into primefaces-extension.js/css).

Still not sure about InputCurrency. As i said, i could commit our company internal InputCurrency because equals to 95% :D

**Comment by ova2** [comment]
@melloware I sent you an email with release instructions.

TabSlideOut is an useful component. In our web shop, we use a similar one (for conract, etc.).

**Comment by melloware** [comment]
@tandraschko I would love if you would upload InputCurrency to PE.  I can take of ShowCase examples for you if you get the component in.

**Comment by Rapster** [comment]
@ova2 here is my e-mail: sebastien.lepage@outlook.com

@melloware would be nice if you could find a prettier tabslideout (and more recent?), as Thomas mentionned, the default theme is not pretty...

**Comment by melloware** [comment]
@Rapster I have looked and I haven't found too many Tab Slide Outs that work besides this one.  I plan on styling it with ThemeRoller like any other component and making it look much prettier.  As far as functionality goes this one is pretty good.  If you find a better one let me know!

Also I am checking in the updated pom's tomorrow.  I was successfully able to update them to do newer Sonatype releases and I was able to successfully perform a SNAPSHOT release today!  So I should be able to take over doing releases from Oleg.  I can help you get set up to do releases as well.
